File: 2-AndroidDesignPatterns/MVC/app/src/main/java/com/example/arch_lab1/fav_products/view/FavProductsActivity.java
package com.example.arch_lab1.fav_products.view;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.arch_lab1.R;
import com.example.arch_lab1.fav_products.controller.Adapter;
import com.example.arch_lab1.fav_products.controller.OnFavClickListener;
import com.example.arch_lab1.model.pojo.Product;
import com.example.arch_lab1.model.repository.Repository;

import java.util.List;

public class FavProductsActivity  extends AppCompatActivity implements OnFavClickListener {
    RecyclerView rv;
    Adapter adapter;
    LinearLayoutManager layoutManager;
    private Repository repository;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.favorite_products);
        rv = findViewById(R.id.favProductsRV);
        layoutManager = new LinearLayoutManager(this);
        rv.setLayoutManager(layoutManager);
        repository=Repository.getInstance(this);
        adapter = new Adapter(this);
        rv.setAdapter(adapter);
        repository.getStoredProducts().observe(this, new Observer<List<Product>>() {
            @Override
            public void onChanged(List<Product> products) {
                adapter.setProducts(products);
            }
        });
    }

    @Override
    public void onUnFavClick(Product product) {
        repository.deleteProduct(product);
    }
}

File: 2-AndroidDesignPatterns/MVVM/app/src/main/java/com/example/mvvm/fav_products/FavProductsActivity.kt
package com.example.mvvm.fav_products

import android.os.Bundle
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.mvvm.databinding.ActivityFavProductsBinding
import com.example.mvvm.model.Product
import com.example.mvvm.model.local.AppDatabase
import com.example.mvvm.model.local.LocalDataSource
import com.example.mvvm.model.remote.RemoteDataSource
import com.example.mvvm.model.repository.ProductRepository

class FavProductsActivity : AppCompatActivity(), com.example.mvvm.fav_products.OnProductClicked {
    private lateinit var binding: ActivityFavProductsBinding
    private lateinit var viewModel: FavViewModel
    private lateinit var adapter: ProductsAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        binding = ActivityFavProductsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        ViewCompat.setOnApplyWindowInsetsListener(binding.main) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        val database = AppDatabase.getDatabase(this)
        val productDao = database.productDao()
        val repository = ProductRepository.getInstance(RemoteDataSource(), LocalDataSource(productDao))

        viewModel = FavViewModel(repository)
        adapter = ProductsAdapter(this)
        binding.recyclerViewFavProducts.layoutManager = LinearLayoutManager(this)
        binding.recyclerViewFavProducts.adapter = adapter

        viewModel.favoriteProducts.observe(this) { products ->
            if (products.isEmpty()) {
                Toast.makeText(this, "No favorite products", Toast.LENGTH_SHORT).show()
            }
            adapter.submitList(products)
        }
    }

    override fun onProductClicked(product: Product) {
        viewModel.deleteProduct(product)
    }
}